### DISCUSSION TITLE
Uvicorn error on startup + reloader not working properly

### DISCUSSION THREAD

**Discussion by illpack** [fastapi/fastapi#6867] [Questions] [answered]
### Problem description

I am running FastAPI inside docker, and the uvicorn reloader is not working as it should. I could not find an similar issue thus I am reporting it here to see if anybode ran into this.
The API **works fine**, but without reloader development work involves shutting down and up docker, plus such an error is probably not fit for production. It still detects changes in the code but only after 1 or 2 minutes, but every time it reloads / starts up it throws the following warning (likely related):

```
api_1  | INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
api_1  | INFO:     Started reloader process [8] using statreload
api_1  | INFO:     Started server process [10]
api_1  | 2021-02-22 10:25:55 | uvicorn.error | {/usr/local/lib/python3.7/site-packages/uvicorn/_impl/asyncio.py:62} | INFO :  Started server process [10]  
api_1  | INFO:     Waiting for application startup.
api_1  | 2021-02-22 10:25:55 | uvicorn.error | {/usr/local/lib/python3.7/site-packages/uvicorn/lifespan/on.py:26} | INFO :  Waiting for application startup.
api_1  | INFO:     Application startup complete.
api_1  | 2021-02-22 10:25:55 | uvicorn.error | {/usr/local/lib/python3.7/site-packages/uvicorn/lifespan/on.py:38} | INFO :  Application startup complete.  
...
api_1  | WARNING:  StatReload detected file change in 'app/utils/colorgram.py'. Reloading...
api_1  | INFO:     Shutting down
api_1  | 2021-02-22 10:30:22 | uvicorn.error | {/usr/local/lib/python3.7/site-packages/uvicorn/_impl/asyncio.py:208} | INFO :  Shutting down
api_1  | INFO:     Waiting for application shutdown.
api_1  | 2021-02-22 10:30:22 | uvicorn.error | {/usr/local/lib/python3.7/site-packages/uvicorn/lifespan/on.py:43} | INFO :  Waiting for application shutdown.
api_1  | INFO:     Application shutdown complete.
api_1  | 2021-02-22 10:30:22 | uvicorn.error | {/usr/local/lib/python3.7/site-packages/uvicorn/lifespan/on.py:46} | INFO :  Application shutdown complete. 
api_1  | INFO:     Finished server process [10]
api_1  | 2021-02-22 10:30:22 | uvicorn.error | {/usr/local/lib/python3.7/site-packages/uvicorn/_impl/asyncio.py:75} | INFO :  Finished server process [10] 
```

My `entrypoint.sh` reads:
```sh
bash -c "uvicorn app.main:app --host 0.0.0.0 --port 8000 --reload"
```

### Environment

* OS: Docker (Linux)
* FastAPI Version: `0.62.0`
* Python version: `3.7.9`
* Uvicorn: `Running uvicorn 0.12.3 with CPython 3.7.9 on Linux`

### Additional context

Dockerfile
```
FROM python:3.7

# Set environment varibles
ENV PYTHONDONTWRITEBYTECODE 1
ENV PYTHONUNBUFFERED 1
ENV CODE_SRC=app

WORKDIR /code/
ENV PATH=$PATH:/code/
ENV PYTHONPATH=$PYTHONPATH:/code/

COPY pyproject.toml poetry.lock ./

RUN set -ex; \
    apt-get update ; \
    apt-get install ffmpeg libsm6 libxext6 -y ; \
    apt-get install jq -y ; \
    apt-get install -y --no-install-recommends git pkg-config ; \
    pip install poetry ; \
    poetry config virtualenvs.create false ; \
    poetry install --no-interaction --no-ansi ; \
    apt-get autoremove -y ; \
    apt-get clean -y ; \
    rm -rf /var/lib/apt/lists/*

COPY . .

ENTRYPOINT [ "./entrypoint.sh"]
````

Thanks!

**Comment by Cleptomania** [comment]
It seems as though you’re copying your directory into the Docker container as opposed to mounting the directory from the host.

When you add files from the host into a container via the Dockerfile it’s not working off of the files on host, but rather compressing those files and uploading them permanently to the container. I don’t know the command you’re using to run the container so I’m not 100% certain this is your issue or not though.

In order to mount the directory from the host you need to add it to your Docker run command like this:

`docker run -v /code:/home/myuser/code my_image`

In the -v argument the directory to the left is the directory inside the container, and the directory to the right is the directory on the host machine. This will map that directory directly into the Docker container, without this the Docker container is unaware of changes to the files on the host machine without a restart.

You can read more about this here https://docs.docker.com/storage/bind-mounts/

**Comment by illpack** [comment]
Hi @Cleptomania thanks for helping out.
I am using docker-compose, where I map the root on the host to `/code` on the Docker:

```yaml
version: "3"

services:

  api:
    build: 
      context: ./app
    volumes:
      - .:/code/
    env_file:
      - ./.env
    ports:
      - "8000:8000"
    command: api
```

The `command` api just calls the `entrypoint.sh` with that argument (this is to account for different envs etc):
```sh

set -e

case "$1" in
  api|uvicorn)
    if [ "$DEVELOPMENT" == "1" ]; then
      echo 'DEVELOPMENT'
      bash -c "uvicorn app.main:app --host 0.0.0.0 --port 8000 --reload"
    else
      echo 'NOT ALLOWED'
    fi
  ;;
esac
```

I believe the mapping is correct, but I might be wrong?
This is my repo structure
```
FASTAPI
├─── .env
├─── .gitignore
├─── docker-compose.yml
├─── README.md
│
└───app
    ├───main.py
    └───Dockerfile
```
thanks!

**Comment by KarlCMB** [comment]
Hi there, I can replicate the problem. Using an older version fixes the issue. (in my case 0.12.3, I didn't try any other)

**Comment by illpack** [comment, marked as answer]
Hi there, I just solved it by explicitly passing [uvicorn ](https://www.uvicorn.org/settings/#development)the path where my development code lives:
```
--reload-dir app
```
so the command becomes
```bash
bash -c "uvicorn app.main:app --host 0.0.0.0 --port 8000 --reload --reload-dir app"
```

Hope this helps!

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @illpack  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
